Graphics.cpp Added!

Paulpoomrit opened the pull request:
We've got most of the functionality working (except for the special filter I'm still working on + main).

Also, I've decided to differ from Shariah's example implementation on two points (I don't know if these are good enough reasons, so feel free to debate, lol).

1) I've made the `PPM` image object parameter to all of the functions in `graphics` a const reference, this means that we'll create a new object within the function itself and return (and being saved later as another separate file) and not modify the original pixel vector in place ( i.e., O(n) space); however, I'd argue the tradeoff is worth it to make the software safer for the client, since that if they mess up they'd still the original copy.

2) follows from that, to prevent dangling reference, we'll make the return type a `const PPM` (i.e., return by value, not by reference). Again, another potential bottleneck when it comes to performance, but better for overall usability (I'd argue here that making it hard for clients to misuse these functions, i.e., provide wrong params and break things halfway, is more important than having 1s faster implementation)

also, here's the command to compile and run:
`clang++ -o program src/Graphics.cpp src/Pixel.cpp src/PPM.cpp -std=c++17` 
`./program`

(in windows i'd assume you'll use g++ instead of clang++?)

Paulpoomrit committed:
Graphics header file added

Paulpoomrit committed:
MakeGreyScale() added

KKangWY committed:
Changed Pixel class to match the requirements

KKangWY committed:
Created and implemented a PPM class that match the requirements

Paulpoomrit committed:
Merge branch 'PPM' into Paul's-branch

Paulpoomrit committed:
Set and get pixels (vector) added

for more flexibility such that we could bulk edit the pixel vector itself

Paulpoomrit committed:
MakeGreyScale done!

Paulpoomrit committed:
Most functionality done!

KKangWY commented:
Here is your requested debate:

1. The client would also have access to functions to copy `PPM` objects and would thus not require the functions in `graphics` if they did need an original copy. However, the filter should still preferably be out-of-place anyways, which this program is, as to not have already filtered pixels affecting unfiltered pixels. The differences in memory required for creating a new object and modifying an existing object would be marginal during runtime because of this; either implementation should be fine.

2. Returning by value and by reference should be at a similar speed, as move is likely to be used in modern versions of C++ and C++ compilers. However, returning a `const PPM` would prevent any modifications to the `PPM` object. Disallowing any changes is not as impactful if copies of the object are always created, but will at the very least prevent a user who wants to change a file's comment from doing so.

Paulpoomrit committed:
Remove const for all returned objs

I don't know why I added this in the first place, lol. Presumably, I thought the main function would only allow one operation being done, which is kinda silly now that I think about it.

remove all const

Update Graphics.cpp

Paulpoomrit committed:
FloydSteinbergDither added

Dithering added

Paulpoomrit committed:
In place emboss filter added

Paulpoomrit committed:
Finalized
